### PR TITLE
Infer schema

### PR DESCRIPTION
**Pull request by Okanmercan99**

### CODE REVIEW

File: tofhir-server/src/main/scala/io/tofhir/server/model/InferTask.scala [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package io.tofhir.server.model
+
+import io.tofhir.engine.model.{DataSourceSettings, SqlSource}
+/**
+ * Infer task instance
```
Comment (dogukan10) [inline tofhir-server/src/main/scala/io/tofhir/server/model/InferTask.scala:5]: Let's explain why we need this model, what is its usage. Moreover, add a comment to indicate that infer functionality works only for SQL sources not others.

File: tofhir-server/src/main/scala/io/tofhir/server/service/SchemaDefinitionService.scala [hunk 2]
```diff
@@ -124,4 +128,67 @@ class SchemaDefinitionService(schemaRepository: ISchemaRepository) extends LazyL
     JArray(rootElement +: elements)
   }
 
+  /**
+   * Infer the schema
+   * @param inferTask infer schema task object
+   * @return
```
Comment (dogukan10) [inline tofhir-server/src/main/scala/io/tofhir/server/service/SchemaDefinitionService.scala:134]: Describe what it returns.

File: tofhir-server/src/main/scala/io/tofhir/server/service/SchemaDefinitionService.scala [hunk 3]
```diff
@@ -124,4 +128,67 @@ class SchemaDefinitionService(schemaRepository: ISchemaRepository) extends LazyL
     JArray(rootElement +: elements)
   }
 
+  /**
+   * Infer the schema
```
Comment (dogukan10) [inline tofhir-server/src/main/scala/io/tofhir/server/service/SchemaDefinitionService.scala:132]: Let's explain the infer functionality here i.e. what it takes, what it does and what it returns at the end.

File: tofhir-server/src/test/scala/io/tofhir/server/project/SchemaEndpointTest.scala [hunk 4]
```diff
@@ -148,13 +186,82 @@ class SchemaEndpointTest extends BaseEndpointTest {
...
+        fieldDefinitions(4).dataTypes.get.head.dataType shouldEqual "integer"
+        fieldDefinitions(5).dataTypes.get.head.dataType shouldEqual "string"
+        fieldDefinitions(6).dataTypes.get.head.dataType shouldEqual "integer"
+      }
+    }
   }
 
   /**
-   * Creates a project to be used in the tests
+   * Creates a project to be used in the tests. Create a table for schema inferring.
```
Comment (dogukan10) [inline tofhir-server/src/test/scala/io/tofhir/server/project/SchemaEndpointTest.scala:228]:
```suggestion
   * Creates a project to be used in the tests. Moreover, it creates a SQL table to be used in the testing of infer schema functionality.
```

File: tofhir-server/src/test/scala/io/tofhir/server/project/SchemaEndpointTest.scala [hunk 5]
```diff
@@ -148,13 +186,82 @@ class SchemaEndpointTest extends BaseEndpointTest {
...
+   * */
+  private def readFileContent(fileName: String): String = {
+    val source: BufferedSource = Source.fromInputStream(getClass.getResourceAsStream(fileName))
+    try source.mkString finally source.close()
+  }
+
+  /**
+   * Run SQL queries for setting up database
+   * */
+  private def runSQL(sql: String): Boolean = {
```
Comment (dogukan10) [inline tofhir-server/src/test/scala/io/tofhir/server/project/SchemaEndpointTest.scala:257]: I think we should close connection and statement at the end?

File: tofhir-server/src/main/scala/io/tofhir/server/service/SchemaDefinitionService.scala [hunk 6]
```diff
@@ -124,4 +128,67 @@ class SchemaDefinitionService(schemaRepository: ISchemaRepository) extends LazyL
...
+   * @param inferTask infer schema task object
+   * @return
+   */
+  def inferSchema(inferTask: InferTask): Future[Option[SchemaDefinition]] = {
+
+    // Execute SQL and get the dataFrame
+    val dataFrame = SourceHandler.readSource("unnamed", SparkConfig.sparkSession,
```
Comment (dogukan10) [inline tofhir-server/src/main/scala/io/tofhir/server/service/SchemaDefinitionService.scala:139]: Does this read the all data or only column metadata ? I think we should add a configuration to read only column metadata for this functionality ?

File: tofhir-server/src/main/scala/io/tofhir/server/service/SchemaDefinitionService.scala [hunk 7]
```diff
@@ -124,4 +128,67 @@ class SchemaDefinitionService(schemaRepository: ISchemaRepository) extends LazyL
...
+      inferTask.sqlSource, inferTask.sourceSettings.head._2, None, None)
+
+    // Create unnamed Schema definition by infer the schema from DataFrame
+    val unnamedSchema = {
+
+      // Map SQL DataTypes to Fhir DataTypes
+      var fieldDefinitions = dataFrame.schema.fields.map(structField => {
+        SimpleStructureDefinition(
+          id = structField.name,
+          path = "unnamed." + structField.name,
```
Comment (dogukan10) [inline tofhir-server/src/main/scala/io/tofhir/server/service/SchemaDefinitionService.scala:149]: Create a constant variable to keep **unnamed** keyword

File: tofhir-server/src/main/scala/io/tofhir/server/service/SchemaDefinitionService.scala [hunk 8]
```diff
@@ -124,4 +128,67 @@ class SchemaDefinitionService(schemaRepository: ISchemaRepository) extends LazyL
...
+          dataTypes = structField.dataType match {
+            case DataTypes.ShortType => Some(Seq(DataTypeWithProfiles(dataType = FHIR_DATA_TYPES.INTEGER, profiles = Some(Seq(s"$FHIR_ROOT_URL_FOR_DEFINITIONS/StructureDefinition/${FHIR_DATA_TYPES.INTEGER}")))))
```
Comment (dogukan10) [inline tofhir-server/src/main/scala/io/tofhir/server/service/SchemaDefinitionService.scala:151]: Let's create a conversion function in SchemaConverter file and use it here

File: tofhir-server/src/main/scala/io/tofhir/server/service/SchemaDefinitionService.scala (hunk 6 above)
Comment (Okanmercan99) [reply tofhir-server/src/main/scala/io/tofhir/server/service/SchemaDefinitionService.scala:139]: I did not find any option that provides this functionality.

Comment (sinaci) [reply tofhir-server/src/main/scala/io/tofhir/server/service/SchemaDefinitionService.scala:139]: From the answers to this SO question (https://stackoverflow.com/questions/44277019/how-to-read-only-n-rows-of-large-csv-file-on-hdfs-using-spark-csv-package), I understand that if we read the data as text and limit, and then apply csv reader on the result, we can achieve what we want during the schema infer process. Can you please check? @Okanmercan99

**Pull request merged**